src/Neo.VM/ReferenceCounterV2.cs

```diff
@@ -0,0 +1,105 @@
+//
+// ReferenceCounterV2.cs file belongs to the neo project and is free
+// software distributed under the MIT software license, see the
+// accompanying file LICENSE in the main directory of the
+// repository or http://www.opensource.org/licenses/mit-license.php
+// for more details.
+//
+// Redistribution and use in source and binary forms with or without
+// modifications are permitted.
+
+using Neo.VM.StronglyConnectedComponents;
+using Neo.VM.Types;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using Array = Neo.VM.Types.Array;
+using Buffer = Neo.VM.Types.Buffer;
+
+namespace Neo.VM
+{
+    /// <summary>
+    /// Used for reference counting of objects in the VM.
+    /// </summary>
+    public sealed class ReferenceCounterV2 : IReferenceCounter
+    {
+        // Keeps the total count of references.
+        private int _referencesCount = 0;
+
+        /// <inheritdoc/>
+        public int Count => _referencesCount;
+
+        /// <inheritdoc/>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool NeedTrack(StackItem item)
+        {
+            throw new NotImplementedException();
+        }
+
+        /// <inheritdoc/>
+        public void AddReference(StackItem item, CompoundType parent)
+        {
+            throw new NotImplementedException();
+        }
+
+        /// <inheritdoc/>
+        public void AddStackReference(StackItem item, int count = 1)
+        {
+            // Increment the reference count by the specified count.
+            _referencesCount += count;
+
+            if (item is CompoundType compoundType)
+            {
+                // Increment the item's stack references by the specified count.
+                compoundType.StackReferences += count;
+                if (compoundType.StackReferences == count)
+                {
+                    foreach (var subItem in compoundType.SubItems)
+                    {
+                        AddStackReference(subItem);
+                    }
+                }
+            }
+        }
+
+        /// <inheritdoc/>
+        public void AddZeroReferred(StackItem item)
+        {
+            throw new NotImplementedException();
+        }
+
+        /// <inheritdoc/>
+        public int CheckZeroReferred()
+        {
+            throw new NotImplementedException();
+        }
+
+        /// <inheritdoc/>
+        public void RemoveReference(StackItem item, CompoundType parent)
+        {
+            throw new NotImplementedException();
+        }
+
+        /// <inheritdoc/>
+        public void RemoveStackReference(StackItem item)
+        {
+            // Increment the reference count by the specified count.
+            _referencesCount--;
+
+            if (item is CompoundType compoundType)
+            {
+                // Increment the item's stack references by the specified count.
+                compoundType.StackReferences--;
```

Comment:
In circular references this counter will become negative

Reply:
exactly, but looks like it wont be a problem.  

for stack item A, B:
 A->C, B->D
 
 then:
 
 C->D, D->C 
 
 
 when we try to remvoe C from A, we will have:
 
 C--; => ref(C) == 0
 D--; => ref(D) == 0 

C--; => ref(C) == -1, calculation stops here, wont keep processing D--;

Reply:
for stack item A, B:
 A->C, B->D => ref.A = 1, ref.B = 1
 

 C->D => ref.C = 1, ref.D = 2.

 D->C => ref.C = 2, ref.D = 2, // ref.D ==2, thus wont process its subitem


 when we try to remvoe C from A, we will have:
 
 C--; => ref(C) == 1

 then, remve D from B, we will have:

D--; => ref.D = 1, // from here the operation related to D and C will be over, the RC will never be decreased anymore.

Reply:
I don't think it can become negative, rather it will have some positive count that is not exactly true (like 1 in RC with no elements on the stack), but this is a safe behavior.